python/lsst/cp/pipe/makeBrighterFatterKernel.py

```diff
@@ -156,1253 +120,308 @@ class MakeBrighterFatterKernelTaskConfig(pexConfig.Config):
-            bfKernel.gainErr = ptcDataset.gainErr
-            bfKernel.noise = ptcDataset.noise
-            bfKernel.noiseErr = ptcDataset.noiseErr
-        except NameError:  # we don't have a ptcDataset to store results from
-            pass
-
-        if self.config.level == 'AMP':
-            bfKernel.ampwiseKernels = kernels
-            ex = self.config.ignoreAmpsForAveraging
-            bfKernel.detectorKernel = bfKernel.makeDetectorKernelFromAmpwiseKernels(detNum, ampsToExclude=ex)
-
-        elif self.config.level == 'DETECTOR':
-            bfKernel.detectorKernel = kernels
-        else:
-            raise RuntimeError('Invalid level for kernel calculation; this should not be possible.')
+            ``outputBfk`` : `lsst.ip.isr.BrighterFatterKernel`
+                Resulting Brighter-Fatter Kernel.
+        """
+        if len(dummy) == 0:
+            self.log.warn("No dummy exposure found.")
 
-        dataRef.put(bfKernel)
+        detector = camera[inputDims['detector']]
+        detName = detector.getName()
 
-        self.log.info('Finished generating kernel(s) for %s' % detNum)
-        return pipeBase.Struct(exitStatus=0)
+        if self.config.level == 'DETECTOR':
+            detectorCorrList = list()
+
+        bfk = BrighterFatterKernel(camera=camera, detectorId=detector.getId(), level=self.config.level)
+        bfk.means = inputPtc.finalMeans  # ADU
+        bfk.variances = inputPtc.finalVars  # ADU^2
+        bfk.rawXcorrs = inputPtc.covariances  # ADU^2
+        bfk.badAmps = inputPtc.badAmps
+        bfk.shape = (inputPtc.covMatrixSide*2 + 1, inputPtc.covMatrixSide*2 + 1)
+        bfk.gain = inputPtc.gain
+        bfk.noise = inputPtc.noise
+        bfk.meanXcorrs = dict()
+        bfk.valid = dict()
 
-    def _applyGains(self, means, xcorrs, ptcData):
-        """Apply the gains calculated by the PtcTask.
+        for amp in detector:
+            ampName = amp.getName()
+            mask = np.array(inputPtc.expIdMask[ampName], dtype=bool)
+
+            gain = bfk.gain[ampName]
+            fluxes = np.array(bfk.means[ampName])[mask]
+            variances = np.array(bfk.variances[ampName])[mask]
+            xCorrList = [np.array(xcorr) for xcorr in bfk.rawXcorrs[ampName]]
+            xCorrList = np.array(xCorrList)[mask]
+
+            if gain <= 0:
+                # We've received very bad data.
+                self.log.warn("Impossible gain recieved from PTC for %s: %f.  Skipping amplifier.",
+                              ampName, gain)
+                bfk.meanXcorrs[ampName] = np.zeros(bfk.shape)
+                bfk.ampKernels[ampName] = np.zeros(bfk.shape)
+                bfk.valid[ampName] = False
+                continue
 
-        It also removes datapoints that were thrown out in the PTC algorithm.
+            fluxes = np.array([flux*gain for flux in fluxes])  # Now in e^-
+            variances = np.array([variance*gain*gain for variance in variances])  # Now in e^2-
```

Comment:
These are the`afwMath.VARIANCECLIP` variances. Do you actually use them in calculations in the code that follows? I only see that you report them in the log info of L239.

Reply:
I only use them in the log I believe.  It was useful to have both to ensure everything was being converted consistently (checking that `mean ~ variance ~ cov(0, 0)`).